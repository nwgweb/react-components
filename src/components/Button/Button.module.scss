@import "../../styles/config";
/**
 * ----------------------------------------
 * Available colors, sizes and modes
 * ----------------------------------------
 */

$colors: (default, primary, transparent, red, blue, green, white);
$sizes: (20, 25, 30, 35, 40, 45, 50);
$modes: (default, pill);
$shadows: ("shadow", "flat");

/**
 * ----------------------------------------
 * Mixins
 * ----------------------------------------
 */

@mixin set-border-radius($mode, $size) {
  @if $mode == pill {
    border-radius: #{$size/2}px;
  } @else {
    border-radius: $border-radius-md;
  }
}

@mixin set-font-size($size) {
  @if $size < 40 {
    font-size: 13px;
  } @else {
    font-size: 16px;
  }
}

/**
 * ----------------------------------------
 * Generate button style
 * ----------------------------------------
 */

@each $color in $colors {
  @each $size in $sizes {
    @each $mode in $modes {
      @each $shadow in $shadows {
        .button-#{$size}-#{$mode}-#{$color}-#{$shadow} {
          position: relative;
          display: inline-block;
          height: #{$size}px;
          line-height: #{$size}px;
          text-align: center;
          font-weight: bold;
          user-select: none;
          border: 0;
          padding: 0 20px;
          transition: all 0.2s ease;

          @if $color == transparent {
            color: $color-text;
            background-color: transparent;
            box-shadow: none !important;
          } @else if $color == white {
            color: $color-text;
          } @else {
            color: $color-white;
          }

          @if $color != transparent {
            @if $shadow == shadow {
              @extend %box-shadow-#{$color};
            }
            @extend %background-color-#{$color};
          }

          @include set-border-radius($mode, $size);
          @include set-font-size($size);

          &:hover {
            text-decoration: none;
            opacity: 0.8;
          }

          &[disabled] {
            background-color: #d4d4d4;
            box-shadow: none;

            &:hover {
              opacity: 1;
            }
          }

          .icon-loading {
            animation: spin;
            animation-duration: 3s;
            animation-iteration-count: infinite;
            animation-timing-function: linear;
          }

          .icon-loading,
          .icon-success,
          .icon-error {
            margin-right: 5px;
            margin-left: -8px;
          }
        }
      }
    }
  }
}

/**
 * ----------------------------------------
 * Loading state
 * ----------------------------------------
 */

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
