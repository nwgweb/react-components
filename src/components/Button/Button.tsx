import React from 'react';
import styles from './Button.module.scss';
import CSSModules from 'react-css-modules';

export type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  size: Number;
  mode?: string;
  color?: string;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  type?: string | undefined;
  shadow?: boolean;
};

const Button: React.FC<ButtonProps> = ({ size, mode = 'default', color = 'default', shadow = true, children, ...rest }) => {
  return (
    <button styleName={`button-${size}-${mode}-${color}-${shadow ? 'shadow' : 'flat'}`} {...rest}>
      {children}
    </button>
  );
};

export default CSSModules(Button, styles);
