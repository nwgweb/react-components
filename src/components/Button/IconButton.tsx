import React from 'react';
import Button from './Button';

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  size: Number;
  mode?: string;
  color?: string;
  iconClass: string;
};

const IconButton: React.FC<ButtonProps> = ({ iconClass, mode, color, size, children, ...rest }) => {
  return (
    <Button size={size} mode={mode} color={color} {...rest}>
      <i className={`fas ${iconClass}`}></i> {children}
    </Button>
  );
};

export default IconButton;
