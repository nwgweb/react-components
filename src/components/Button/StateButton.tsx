import React, { useState, useEffect } from 'react';
import styles from './Button.module.scss';
import CSSModules from 'react-css-modules';
import Button, { ButtonProps } from './Button';

type StateButtonProps = ButtonProps & {
  isLoading?: boolean;
  isSuccess?: boolean;
  isError?: boolean;
};

const StateButton: React.FC<StateButtonProps> = ({ isLoading, isSuccess, isError, children, ...rest }) => {
  const [showLoader, setShowLoader] = useState(isLoading);

  // Add delay to loading state change
  useEffect(() => {
    if (isLoading) {
      setShowLoader(true);
    }

    if (!isLoading && showLoader) {
      const timeout = setTimeout(() => {
        setShowLoader(false);
      }, 300);

      return () => {
        clearTimeout(timeout);
      };
    }
  }, [showLoader, isLoading, isError]);

  return (
    <Button {...rest} color={`${isError ? 'red' : rest.color ? rest.color : 'default'}`}>
      {showLoader && !isError && <i styleName="icon-loading" className="fas fa-circle-notch"></i>}
      {isSuccess && !showLoader && <i styleName="icon-success" className="fas fa-check-circle"></i>}
      {isError && <i styleName="icon-error" className="fas fa-exclamation-circle"></i>}
      {isError ? <>Error</> : children}
    </Button>
  );
};

export default CSSModules(StateButton, styles);
