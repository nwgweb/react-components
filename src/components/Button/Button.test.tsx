import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';

describe('Button', () => {
  test('snapshot renders', () => {
    const component = renderer.create(<Button size={35}></Button>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
