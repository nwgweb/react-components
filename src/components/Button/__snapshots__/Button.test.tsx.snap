// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot renders 1`] = `
<button
  className="button-35-default-default"
/>
`;
